ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java

```diff
@@ -0,0 +1,132 @@
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.net.URISyntaxException;
+
+/**
+ * <p>ManagedStreamingIngestClient</p>
+ * <p>
+ * This class combines a managed streaming client with a queued streaming client, to create an optimized experience.
+ * Since the streaming client is communicating directly with the engine, it's more prone to failure, so this class
+ * holds both a streaming client and a queued client.
+ * At first it tries using the streaming client, and if it fails retries it {@value MAX_RETRY_CALLS} times, and if that
+ * still fails it then uses the queued streaming client.
+ */
+public class ManagedStreamingIngestClient implements IngestClient {
+
+    private final static Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+    public static final int MAX_RETRY_CALLS = 10;
+    private final QueuedIngestClient queuedIngestClient;
+    private final StreamingIngestClient streamingIngestClient;
+
+    public ManagedStreamingIngestClient(ConnectionStringBuilder dmConnectionStringBuilder,
+                                        ConnectionStringBuilder engineConnectionStringBuilder) throws URISyntaxException {
+        log.info("Creating a new ManagedStreamingIngestClient from connection strings");
+        queuedIngestClient = new QueuedIngestClient(dmConnectionStringBuilder);
+        streamingIngestClient = new StreamingIngestClient(engineConnectionStringBuilder);
+    }
+
+    public ManagedStreamingIngestClient(ResourceManager resourceManager,
+                                        AzureStorageClient storageClient,
+                                        StreamingClient streamingClient) {
+        log.info("Creating a new ManagedStreamingIngestClient from raw parts");
+        queuedIngestClient = new QueuedIngestClient(resourceManager, storageClient);
+        streamingIngestClient = new StreamingIngestClient(streamingClient);
+    }
+
+    @Override
+    public IngestionResult ingestFromFile(FileSourceInfo fileSourceInfo, IngestionProperties ingestionProperties) throws IngestionClientException, IngestionServiceException {
+        Ensure.argIsNotNull(fileSourceInfo, "fileSourceInfo");
+        Ensure.argIsNotNull(ingestionProperties, "ingestionProperties");
+
+        fileSourceInfo.validate();
+        ingestionProperties.validate();
+        try {
+            StreamSourceInfo streamSourceInfo = IngestionUtils.fileToStream(fileSourceInfo);
+            return ingestFromStream(streamSourceInfo, ingestionProperties);
+        } catch (FileNotFoundException e) {
+            log.error("File not found when ingesting a file.", e);
+            throw new IngestionClientException("IO exception - check file path.", e);
+        }
+
+    }
+
+    @Override
+    public IngestionResult ingestFromBlob(BlobSourceInfo blobSourceInfo, IngestionProperties ingestionProperties) throws IngestionClientException, IngestionServiceException {
```

Comment:
I suggest documenting the fact that this method doesn't follow the class behavior (as stated in its documentation)

Reply:
Do you think I also should document it on the method, or just the class?